tool/tsh/proxy.go

```diff
@@ -255,6 +258,91 @@ func mkLocalProxyCerts(certFile, keyFile string) ([]tls.Certificate, error) {
 	return []tls.Certificate{cert}, nil
 }
 
+func onProxyCommandAPP(cf *CLIConf) error {
+	tc, err := makeClient(cf, false)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	appCerts, err := loadAppCertificate(tc, cf.AppName)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	address, err := utils.ParseAddr(tc.WebProxyAddr)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	addr := "localhost:0"
+	if cf.LocalProxyPort != "" {
+		addr = fmt.Sprintf("127.0.0.1:%s", cf.LocalProxyPort)
+	}
+
+	listener, err := net.Listen("tcp", addr)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	lp, err := alpnproxy.NewLocalProxy(alpnproxy.LocalProxyConfig{
+		Listener:           listener,
+		RemoteProxyAddr:    tc.WebProxyAddr,
+		Protocol:           alpncommon.ProtocolHTTP,
+		InsecureSkipVerify: cf.InsecureSkipVerify,
+		ParentContext:      cf.Context,
+		SNI:                address.Host(),
+		AWSCredentials:     nil,
+		Certs:              []tls.Certificate{appCerts},
+	})
+	if err != nil {
+		if cerr := listener.Close(); cerr != nil {
+			return trace.NewAggregate(err, cerr)
+		}
+		return trace.Wrap(err)
+	}
+
+	defer lp.Close()
+	go func() {
+		if err = lp.Start(cf.Context); err != nil {
+			log.WithError(err).Errorf("Failed to start local proxy.")
+		}
+	}()
+
+	fmt.Printf("Proxying connections to %s on %v\n", cf.AppName, lp.GetAddr())
+
+	<-cf.Context.Done()
+
+	return nil
+}
+
+func loadAppCertificate(tc *client.TeleportClient, appName string) (tls.Certificate, error) {
+	key, err := tc.LocalAgent().GetKey(tc.SiteName, client.WithAppCerts{})
+	if err != nil {
+		return tls.Certificate{}, trace.Wrap(err)
+	}
+	cc, ok := key.AppTLSCerts[appName]
+	if !ok {
+		return tls.Certificate{}, trace.NotFound("please login into the application first. 'tsh app login'")
+	}
+	cert, err := tls.X509KeyPair(cc, key.Priv)
+	if err != nil {
+		return tls.Certificate{}, trace.Wrap(err)
+	}
+	if len(cert.Certificate) < 1 {
+		return tls.Certificate{}, trace.NotFound("invalid certificate length")
```

Comment:
I agree with @ravicious. While this scenario may not be very likely, the literal error message might confuse most users. Could we change this to something that is a bit more descriptive/actionable so that a user may be able to potentially resolve this on their own instead of having to post in the community slack or look at the code?
```suggestion
		return tls.Certificate{}, trace.NotFound("invalid certificate - please login into the application again. 'tsh app login'")
```

Reply:
fixed in https://github.com/gravitational/teleport/pull/10509/commits/0681c26fa092aa600631cfb77d4d756a7da3b94a